scripts/saml-configuration

```diff
@@ -0,0 +1,174 @@
+import logging
+import os
+import pwd
+import sys
+from typing import Any, Dict, Iterator, List, Tuple
+
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+sys.path.append(BASE_DIR)
+# Enter the Zulip virtual environment.
+from scripts.lib.setup_path import setup_path
+
+setup_path()
+
+# Settings for django.conf
+os.environ["DJANGO_SETTINGS_MODULE"] = "zproject.settings"
+
+from cryptography import x509
+from cryptography.hazmat.backends import default_backend
+from django.conf import settings
+from onelogin.saml2.idp_metadata_parser import OneLogin_Saml2_IdPMetadataParser as Parser
+from onelogin.saml2.utils import OneLogin_Saml2_Utils as Utils
+
+"""A script to manage some of the SAML configuration automatically.
+
+Current status:
+Keep the signing certificates of the IdPs listed in
+settings.SOCIAL_AUTH_SAML_ENABLED_IDPS up to date.
+"""
+
+
+# The template for the path to the IdP certificate.
+idp_cert_path_template: str = "/etc/zulip/saml/idps/{}.crt"
+# The format string used for logging.
+log_format: str = "%(levelname)s:%(asctime)s saml-configuration: %(message)s"
+
+
+def is_valid(cert: x509.Certificate) -> bool:
+    """Check if the given certificate is currently valid."""
+    now: datetime.datetime = datetime.datetime.now()
+    return now >= cert.not_valid_before and now <= cert.not_valid_after
+
+
+def public_cert_open(cert_path: str, fix_permissions: bool = True) -> io.TextIOWrapper:
+    """Open/create a file for a public certificate.
+
+    If the file needs to be created, its permissions are set to
+    `-rw-r--r--` (0644). If the file already exists, the permissions
+    will only be changed if 'fix_permissions' is true (the default).
+    If the current umask value is more restrictive, it dominates the
+    new permissions.
+
+    Return an open file object. May raise an OSError, for example, if
+    the file already exists and is a directory.
+    """
+    flags: int = os.O_WRONLY | os.O_CREAT | os.O_TRUNC
+    mode: int = 0o644
+
+    if os.path.isfile(cert_path) and os.stat(cert_path).st_mode != mode:
+        if fix_permissions:
+            os.chmod(cert_path, mode)
+        else:
+            logging.warning("file '%s' does not have mode %o", mode)
+
+    # At this point, the mode of the file will stay unchanged if it
+    # already exists.
+    return os.fdopen(os.open(cert_path, flags, mode), mode="w")
+
+
+parser: argparse.ArgumentParser = argparse.ArgumentParser()
+parser.add_argument("--debug", action="store_true", help="enable debug logging")
+parser.add_argument(
+    "--fix-permissions", action="store_true", help="fix inappropriate certificate file permissions"
+)
+args: argparse.Namespace = parser.parse_args()
+
+if args.debug:
+    logging.basicConfig(format=log_format, level=logging.DEBUG)
+else:
+    logging.basicConfig(format=log_format)
+
+if pwd.getpwuid(os.getuid()).pw_name != "zulip":
+    logging.critical("This script needs to be executed as user 'zulip'")
+    sys.exit(1)
```

Comment:
Exceptions would become `raise CommandError` if we make this a management command, but in this case I think we just get to remove the check, since the manage.py tooling takes care of it for us.